cranelift/jit/src/compiled_blob.rs

```diff
@@ -93,6 +93,31 @@ impl CompiledBlob {
                     let imm26 = (diff as u32) << chop >> chop;
                     unsafe { modify_inst32(iptr, |inst| inst | imm26) };
                 }
+                Reloc::Aarch64AdrGotPage21 => {
+                    // Set the immediate value of an ADRP to bits [32:12] of X; check that –2^32 <= X < 2^32
+                    let base = get_got_entry(name);
+                    let what = unsafe { base.offset(isize::try_from(addend).unwrap()) };
```

Comment:
We only emit `Aarch64AdrGotPage21` relocations with an `addend` of `0`. Would you mind removing the offset, and instead asserting that it's `0`?

Additionally, to fully implement the operation `Page(G(GDAT(S+A)))-Page(P)` we would need to modify the function `get_got_entry` that's passed in, as we would need to make the offset calculation before looking up in the GOT. Asserting that the `addend` is zero means that we can avoid that refactoring.

Reply:
Good catch! :+1: